### DISCUSSION TITLE
[Feature request] Support Influxdb sub queries in the query editor

### DISCUSSION THREAD

**Discussion by svitlanacs** [grafana/grafana#43847] [Observability]
Subqueries have recently been implemented in InfluxDB, however, I was not able to find how to write up a subquery in Grafana.

Here's the most promising approach I have tried:
![screen shot 2017-02-28 at 5 56 20 pm](https://cloud.githubusercontent.com/assets/26101874/23442762/71f3066a-fddf-11e6-80ff-0981e673f97b.png)

However, the problem is that InfluxDB wants its contents in double quotes and Grafana wraps the contents of FROM field with double quotes as well, so the resulting query fails to executed:
Request:
SELECT mean("sum") / 100 FROM **"**(SELECT sum(**"**count")/60 FROM "read_bytes" WHERE time > XXms and time < YYms GROUP BY time(1m))" WHERE time > XXms and time < YYms GROUP BY time(5m) fill(0);

Response error complaining about test after double quotes:
{
 "error": "error parsing query: found )/60 FROM , expected ; at line 1, char 46",
 "message": "error parsing query: found )/60 FROM , expected ; at line 1, char 46"
}

What is the correct approach to write subqueries in Grafana against InfluxDB?

**Comment by bergquist** [comment]
You have to use raw queries to support subqueries.

Click the small hamburger menu in the metric raw to toggle between rawqueries and the query editor
![image](https://cloud.githubusercontent.com/assets/618863/23452700/90bc0412-fe65-11e6-856a-bba24dc75215.png)


http://play.grafana.org/dashboard/db/influxdb-raw-query-template-var?panelId=1&fullscreen&edit

I changed this issue to feature request since its new functionality that doesn't exist in Grafana.

**Comment by svitlanacs** [comment]
Thank you very much, @bergquist . I got unblocked by writing out the raw query in the query editor.

But it would still be nice to be able to write subqueries not in the raw mode but in the nice Grafana interface, so I will leave this issue open.

**Comment by smalenfant** [comment]
+1 here. Been working for me with raw queries from InfluxDB but this could be a great enhancement.

**Comment by lewg** [comment]
Although you can put in the raw query, it's been my experience thus far that you can't setup alerts on the resulting graph.

**Comment by ghost** [comment]
+1

**Comment by spali** [comment]
design idea:
because subqueries could be recursive and I don't see a easy and clean way for the GUI if this is build inside one query. Why not allow to use i.e. Query "A" to be referenced in another query as measurement.

I try to avoid the raw queries as much as possible, but I have a lot of peak detection in graphs that use mean to still get the peaks of the flattened data by `mean`.

## current way
#### Query A (can be done with query builder):
```
SELECT mean("Percent_User_Time") AS "User"
  FROM "win_cpu" WHERE ("host" =~ /^$hostname$/ AND "instance" = '_Total' AND "objectname" = 'Processor') AND $timeFilter GROUP BY time($__interval)
```
#### Query B (must be done in raw mode):
```
select "Percent_User_Time_MAX" from (
select max("Percent_User_Time") as "Percent_User_Time_MAX",
            mean("Percent_User_Time") as "Percent_User_Time_MEAN"
 FROM "win_cpu" WHERE ("host" =~ /^$hostname$/ AND "instance" = '_Total' AND "objectname" = 'Processor') AND $timeFilter
 GROUP BY time($__interval)
) where "Percent_User_Time_MAX" != "Percent_User_Time_MEAN"
```

## with the possibility to reference other queries I could clean this up with the query builder:
#### Query A (can be done with query builder):
```
SELECT mean("Percent_User_Time") AS "User"
  FROM "win_cpu" WHERE ("host" =~ /^$hostname$/ AND "instance" = '_Total' AND "objectname" = 'Processor') AND $timeFilter GROUP BY time($__interval)
```
#### Query B (could be done with query builder as disabled query):
```
select max("Percent_User_Time") as "Percent_User_Time_MAX",
            mean("Percent_User_Time") as "Percent_User_Time_MEAN"
 FROM "win_cpu" WHERE ("host" =~ /^$hostname$/ AND "instance" = '_Total' AND "objectname" = 'Processor') AND $timeFilter
 GROUP BY time($__interval)
```

#### Query C (could be done with query builder):
```
select "Percent_User_Time_MAX" from QUERY_B where "Percent_User_Time_MAX" != "Percent_User_Time_MEAN"
```

even when I first want to sum up `Percent_User_Time` with `Percent_Privileged_Time` and  `Percent_Interrupt_Time` as total and work with this or reuse this in the query with the details, I could just reference the "total" query

didn't looked at the code, but this is probalby just a replace of the QUERY_X tag with the (QUERY) before sending the queries to influxdb

**Comment by ximply** [comment]
I use toggle edit mode but graph show nothing, then in influxdb console is ok,
SELECT top(t,15),protocol FROM (SELECT sum("total") AS t FROM "rp_two_month"."app_response" WHERE time > now() - 1h GROUP BY "project", "cluster", "model", "protocol", time(10m) fill(0))
name: app_response
time                 top   protocol
----                 ---   --------
2019-05-06T06:50:00Z 9470  CmdCache/Hint
2019-05-06T06:50:00Z 8389  Pprotocol/a
2019-05-06T06:50:00Z 7849  DBC/Song
2019-05-06T07:00:00Z 12050 Dp/a
2019-05-06T07:00:00Z 10817 Pp/a
2019-05-06T07:00:00Z 9821  CmdCache/Hint
2019-05-06T07:10:00Z 13419 Pp/i
2019-05-06T07:10:00Z 8598  DBC/Song
2019-05-06T07:10:00Z 8155  Dp/i
2019-05-06T07:10:00Z 8058  CmdCache/Hint
2019-05-06T07:20:00Z 12137 CmdCache/Hint
2019-05-06T07:20:00Z 11115 Pp/x
2019-05-06T07:20:00Z 10188 Dp/f
2019-05-06T07:30:00Z 9932  Dp/a
2019-05-06T07:30:00Z 8753  Pp/a

**Comment by ximply** [comment]
query inspector is ok too.

**Comment by HBLiB** [comment]
in the latest version its not possible anymore,
in 6.0.2 its possible and loving it, but in 6.1.6 its broken... :'(

So if you need this, downgrade to 6.0.2. And the render option also shows some "issues"...

**Comment by voiprodrigo** [comment]
ability to reference another query in the select would be great addition!

**Comment by wfjm** [comment]
+1.  
sub-queries are essential in several situations.  
Especially because there is no Flux query builder, so using InfluxQL is the more convenient way.

**Comment by gabor** [comment]
we are moving feature-requests to github-discussions from github-issues.

**Comment by itsmylife** [comment]
Thank you for submitting a feature request. We’ve reviewed your request and have decided not to implement this feature for the following reason:
- It will bring a huge maintenance toll as each query needs to be parsed and serialized
- Implementation will be quite messy
- Currently it has a low priority
- We would like to run a UX research and have a design doc first

Even when we make the decision not to pursue a feature, the submission encourages conversations and ideas within our development team to continue making improvements, so please keep sharing!
